### PR TITLE
BrushUp: add scroll shadow

### PR DESCRIPTION
**Pull request by cguglielmo**
The scroll shadow visualizes whether there is scrollable content.
It is installed automatically when scrollbars are installed.
The shadow is a div that is placed after the scrollable.
The div has 4 box shadows, one for each side. Since it is placed
after the scrollable, it has to track the style, visibility and position
of the scrollable and adjust itself accordingly. The size and position
are mainly updated by the already existing scrollbars.update calls by
the layouts. Additionally, dom mutation and intersection observers are
used to track dom movements and visibility changes.

Some alternative solutions that did not work well:
- Using a box shadow inside the scrollable container: The box shadow
  lays under the elements, so if they have a background color the shadow
  won't be visible completely.
- Using position sticky: only works with relative positioning and does
  not work well for horizontal scrolling
- Using position absolute and update position while scrolling as done
  for the scrollbars itself: does not work well with native scrolling
  because scroll events are not fired immediately -> shadow jumps
  around.

292522

### CODE REVIEW

**Comment by cguglielmo** [comment]
run tests

**Comment by avandorp** [comment]
run tests

**Comment by cguglielmo** [comment]
run tests

**Comment by avandorp** [comment]
run tests

File: eclipse-scout-core/src/form/fields/groupbox/GroupBox.js [hunk 1]
```diff
@@ -360,6 +364,36 @@ export default class GroupBox extends CompositeField {
...
+  _updateScrollShadow() {
+    if (!this.rendered) {
+      return;
+    }
+    let hasScrollShadowTop = this.hasScrollShadow('top');
+    let hasScrollShadowBottom = this.hasScrollShadow('bottom');
+    let oldHasScrollShadowTop = this.$container.hasClass('has-scroll-shadow-top');
+    this.$container.toggleClass('has-scroll-shadow-top', hasScrollShadowTop);
+    this.$container.toggleClass('has-scroll-shadow-bottom', hasScrollShadowBottom);
+    if (oldHasScrollShadowTop !== hasScrollShadowTop) {
```
Comment (mvilliger) [inline eclipse-scout-core/src/form/fields/groupbox/GroupBox.js:381]: There is no need to compare the bottom scroll-shadow as well?

File: eclipse-scout-core/src/scrollbar/Scrollbar.js [hunk 2]
```diff
@@ -76,6 +76,9 @@ export default class Scrollbar extends Widget {
     this._dirReverse = this.axis === 'x' ? 'right' : 'bottom';
     this._scrollDir = this.axis === 'x' ? 'scrollLeft' : 'scrollTop';
 
+    // this.$parent.appendDiv('invisible scroll-shadow ' + this._dir);
+    // this.$parent.appendDiv('invisible scroll-shadow ' + this._dirReverse);
```
Comment (mvilliger) [inline eclipse-scout-core/src/scrollbar/Scrollbar.js:80]: still required?

File: eclipse-scout-core/src/scrollbar/scrollbars.js [hunk 3]
```diff
@@ -127,6 +140,190 @@ export function _installNativeInternal($container, options) {
...
+  _installIntersectionObserver();
+  intersectionObserver.observe($container[0]);
+}
+
+function uninstallScrollShadow($container, session) {
+  let $shadow = $container.data('scroll-shadow');
+  if ($shadow) {
+    $shadow.remove();
+  }
+  intersectionObserver.unobserve($container[0]);
```
Comment (mvilliger) [inline eclipse-scout-core/src/scrollbar/scrollbars.js:166]: remove 'scroll-shadow' and 'scroll-shadow-style' from $container data?
And remove 'scroll' event listener from container?

File: eclipse-scout-core/src/widget/widgets.js [hunk 4]
```diff
@@ -67,10 +67,26 @@ export function findFirstFocusableWidget(widgets, container, checkTabbable) {
   return arrays.find(widgets, widget => widget.isFocusable(checkTabbable));
 }
 
+export function preserveAndSetProperty(setter, getter, preserver, preserverName) {
```
Comment (mvilliger) [inline eclipse-scout-core/src/widget/widgets.js:70]: Add jsdoc

**Pull request closed without merging**